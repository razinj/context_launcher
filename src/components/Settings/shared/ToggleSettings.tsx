import React, { PropsWithChildren, useState } from 'react'
import { Pressable, StyleSheet, View } from 'react-native'
import CustomIcon from '../../shared/CustomIcon'
import SettingsItemLabel from './SettingsItemLabel'
import { settingItemButtonRippleConfig } from './values'

export type ToggleSettingsProps = PropsWithChildren<{ title: string; description?: string }>

const ToggleSettings = ({ title, description, children }: ToggleSettingsProps) => {
  const [renderSettings, setRenderSettings] = useState(false)

  return (
    <>
      <View style={renderSettings ? [styles.wrapper, styles.activeWrapper] : styles.wrapper}>
        <Pressable
          testID='toggle-settings-button'
          onPress={() => setRenderSettings(!renderSettings)}
          android_disableSound={true}
          android_ripple={settingItemButtonRippleConfig}
          style={styles.pressable}>
          <SettingsItemLabel
            title={title}
            description={description}
            titleStyle={renderSettings ? [styles.activeTitle] : []}
          />
          <CustomIcon name={`chevron-${renderSettings ? 'up' : 'down'}`} size={30} color='#808080' />
        </Pressable>
      </View>

      {renderSettings && children}
    </>
  )
}

const styles = StyleSheet.create({
  wrapper: {
    borderRadius: 5,
  },
  activeWrapper: {
    backgroundColor: '#ececec',
  },
  pressable: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  activeTitle: {
    fontWeight: 'bold',
  },
})

export default ToggleSettings
