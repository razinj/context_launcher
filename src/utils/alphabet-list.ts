import { AppDetails } from '../models/app-details'
import { AppLetterIndex } from '../models/list-letter-index'

type LetterMap = {
  [key: string]: number
}

const alphabet = [
  'A',
  'B',
  'C',
  'D',
  'E',
  'F',
  'G',
  'H',
  'I',
  'J',
  'K',
  'L',
  'M',
  'N',
  'O',
  'P',
  'Q',
  'R',
  'S',
  'T',
  'U',
  'V',
  'W',
  'X',
  'Y',
  'Z',
]

/**
 * @returns Indexed alphabet letters in an object where the key is the alphabet and the value is the index
 */
export const getLettersMap = (): LetterMap => {
  const letterMap: LetterMap = {}

  alphabet.forEach((letter: string, index: number) => (letterMap[letter] = index + 1))

  return letterMap
}

/**
 * @param value Any string value
 * @returns A letter if the first chareceter of the value is an alphabet, otherwise it returns a '#'
 */
export const getFirstLetter = (value: string): string => {
  const firstChar = value.trim().substring(0, 1).toUpperCase()
  const isValidLetter = getLettersMap()[firstChar]

  if (isValidLetter) return firstChar

  return '#'
}

export const getAppsLetterIndex = (apps: AppDetails[]): AppLetterIndex[] => {
  const appsLetterIndex: AppLetterIndex[] = []
  const treatedLetters: string[] = []

  apps.forEach(({ name }: AppDetails, index: number) => {
    const letter = getFirstLetter(name)

    if (treatedLetters.includes(letter)) return

    appsLetterIndex.push({ letter, index })
    treatedLetters.push(letter)
  })

  return appsLetterIndex
}
