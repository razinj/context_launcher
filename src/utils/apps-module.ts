// Native modules
import AppsModule from '../native-modules/AppsModule'

export const launchApp = (packageName: string) => AppsModule.launchApplication(packageName)

export const showAppDetails = (packageName: string) => AppsModule.showApplicationDetails(packageName)

export const requestAppUninstall = (packageName: string) => AppsModule.requestApplicationUninstall(packageName)
