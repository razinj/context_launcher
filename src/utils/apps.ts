// Models
import { AppDetails, AppDetailsWithIcon } from '../models/app-details'
import { PinnedApp } from '../models/pinned-app'

export const getAppsByLabel = (apps: AppDetails[], query: string): AppDetails[] => {
  return apps.filter((app: AppDetails) => app.label.match(new RegExp(query, 'gi')))
}

export const getPinnedAppByName = (pinnedApps: PinnedApp[], appDetails: AppDetailsWithIcon): PinnedApp => {
  const pinnedApp = pinnedApps.find(({ name }: PinnedApp) => name === appDetails.name)

  if (pinnedApp) return pinnedApp

  return { ...appDetails, isTemporary: false, isPermanent: false }
}
