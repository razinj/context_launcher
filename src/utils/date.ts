export const getTimeFromDate = (stringDate?: string): string => {
  if (!stringDate) return 'unset'

  const date = new Date(stringDate)

  const singleDigitHours = `${date.getHours()}`.length == 1
  const singleDigitMinutes = `${date.getMinutes()}`.length == 1

  const hours = singleDigitHours ? `0${date.getHours()}` : date.getHours()
  const minutes = singleDigitMinutes ? `0${date.getMinutes()}` : date.getMinutes()

  return `${hours}:${minutes}`
}

export const getDateFromStringWithCurrentDateValue = (stringDate?: string): Date | undefined => {
  if (!stringDate) return

  const currentDate = new Date()
  const fullDate = new Date(stringDate)
  return new Date(
    currentDate.getFullYear(),
    currentDate.getMonth(),
    currentDate.getDate(),
    fullDate.getHours(),
    fullDate.getMinutes(),
    0,
    0
  )
}

export const stripDateFromSeconds = (date: Date): Date => {
  return new Date(date.getFullYear(), date.getMonth(), date.getDate(), date.getHours(), date.getMinutes(), 0, 0)
}
