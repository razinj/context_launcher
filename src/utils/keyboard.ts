import { Keyboard } from 'react-native'

export const dismissKeyboard = () => Keyboard.dismiss()
