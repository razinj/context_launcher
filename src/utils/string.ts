export const truncateString = (value: string | undefined, maxLength: number): string | undefined => {
  return value && value.length > maxLength ? `${value.substring(0, maxLength)}...` : value
}
