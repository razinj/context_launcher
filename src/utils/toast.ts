import { ToastAndroid } from 'react-native'

export const displayToast = (message: string, duration: number = ToastAndroid.SHORT) => {
  ToastAndroid.show(message, duration)
}
