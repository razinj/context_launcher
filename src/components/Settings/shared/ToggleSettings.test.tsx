import { fireEvent, screen, waitFor } from '@testing-library/react-native'
import React from 'react'
import { Text } from 'react-native'
import { renderWithProvider } from '../../../../utils/test/utils'
import ToggleSettings, { ToggleSettingsProps } from '../shared/ToggleSettings'

describe('<ToggleSettings /> Tests', () => {
  const requiredProps: ToggleSettingsProps = {
    title: 'An item label',
  }

  it('should render correctly and match snapshot - required props', () => {
    renderWithProvider(
      <ToggleSettings {...requiredProps}>
        <Text>A Child</Text>
      </ToggleSettings>
    )

    expect(screen.toJSON()).toMatchSnapshot()
    expect(screen.getByTestId('toggle-settings-button')).toBeOnTheScreen()
    expect(screen.getByText(requiredProps.title)).toBeOnTheScreen()
    expect(screen.queryByText('A Child')).not.toBeOnTheScreen()
  })

  it('should render correctly and match snapshot - required and optional props', () => {
    const allProps: ToggleSettingsProps = {
      ...requiredProps,
      description: 'An item description',
    }

    renderWithProvider(
      <ToggleSettings {...allProps}>
        <Text>A Child</Text>
      </ToggleSettings>
    )

    expect(screen.toJSON()).toMatchSnapshot()
    expect(screen.getByTestId('toggle-settings-button')).toBeOnTheScreen()
    expect(screen.getByText(allProps.title)).toBeOnTheScreen()
    expect(screen.getByText(allProps.description!)).toBeOnTheScreen()
    expect(screen.queryByText('A Child')).not.toBeOnTheScreen()
  })

  it('should display children when toggled', () => {
    renderWithProvider(
      <ToggleSettings {...requiredProps}>
        <Text>A Child</Text>
      </ToggleSettings>
    )

    const toggleButton = screen.getByTestId('toggle-settings-button')

    expect(toggleButton).toBeOnTheScreen()

    fireEvent.press(toggleButton)

    waitFor(() => {
      expect(screen.queryByText('A Child')).toBeOnTheScreen()
    })
  })
})
