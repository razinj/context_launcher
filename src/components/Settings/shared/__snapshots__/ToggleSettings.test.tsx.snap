// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToggleSettings /> Tests should render correctly and match snapshot - required and optional props 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      style={
        {
          "borderRadius": 5,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
        testID="toggle-settings-button"
      >
        <View
          style={
            [
              {
                "padding": 10,
              },
              undefined,
            ]
          }
          testID="wrapper"
        >
          <Text
            style={
              [
                {
                  "color": "#808080",
                },
                [],
              ]
            }
            testID="title"
          >
            An item label
          </Text>
          <Text
            style={
              {
                "color": "#ccc",
              }
            }
          >
            An item description
          </Text>
        </View>
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            [
              {
                "color": "#808080",
                "fontSize": 30,
              },
              undefined,
              {
                "fontFamily": "Material Design Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              {},
            ]
          }
        >
          󰅀
        </Text>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;

exports[`<ToggleSettings /> Tests should render correctly and match snapshot - required props 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      style={
        {
          "borderRadius": 5,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
        testID="toggle-settings-button"
      >
        <View
          style={
            [
              {
                "padding": 10,
              },
              undefined,
            ]
          }
          testID="wrapper"
        >
          <Text
            style={
              [
                {
                  "color": "#808080",
                },
                [],
              ]
            }
            testID="title"
          >
            An item label
          </Text>
        </View>
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            [
              {
                "color": "#808080",
                "fontSize": 30,
              },
              undefined,
              {
                "fontFamily": "Material Design Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              {},
            ]
          }
        >
          󰅀
        </Text>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
